File: other_code/python_plus_numpy/fastestPossible.py
import numpy as np
import random
import time

NNODES = 100
NRUNS  = 100
def do_work():

    max_len = 0
    max_filled = -1

    # Run 100 walks on the path
    for i in range(NNODES):
        loc = 0
        length = 0
        while loc <= max_filled:
            if loc == 0:
                loc += 1
            else:
                loc = loc + 1 if random.random() > 0.5 else loc - 1
            length += 1
        max_filled = loc
        if length > max_len:
            max_len = length

    return max_len

if __name__ == "__main__":

    t0 = time.time()
    lengths = []
    for i in range(NRUNS):
        if i % 10 == 0: print(i)
        lengths.append(do_work())
    t1 = time.time()

    print("max", np.max(lengths))
    print("min", np.min(lengths))
    print("ave", np.mean(lengths))
    print("runtime", t1-t0)

File: cpp/fastestPossible.cpp
// (c) Nikolaus Howe 2018

#include <iostream>
#include <cstdlib>
#include <time.h>

// An optimized version (but not using any approximations)
// of the standard random walk on the path,
// tarting from the leftmost node.

#define NNODES 100
#define NRUNS  100

int doWork();
int getMax(int *a);
int getMin(int *a);

int main() {
    clock_t start = clock();
    int lengths[NRUNS];
    int sum = 0;
    for (int i = 0; i < NRUNS; i++) {
        if (i % 10 == 0) std::cout << i << std::endl;
        int res = doWork();
        lengths[i] = res;
        sum += res;
    }

    clock_t end = clock();

    std::cout << "max " << getMax(lengths) << std::endl;
    std::cout << "min " << getMin(lengths) << std::endl;
    std::cout << "ave " << sum / NRUNS << std::endl;
    std::cout << "runtime " << (double) (end-start) / CLOCKS_PER_SEC << std::endl;
}

int doWork() {

    int max_len = 0;
    int max_filled = -1;
    
    // Run NNODES walks on the path
    for (int i = 0; i < NNODES; i++) {
        int loc = 0;
        int length = 0;
        while (loc <= max_filled) {
            if (loc == 0) {
                loc++;
            } else {
                //std::srand(std::time(nullptr));
                double rand = std::rand();
                //std::cout << rand << std::endl;
                if (rand > RAND_MAX / 2) {
                    loc++;
                } else {
                    loc--;
                }
            }
            length++;
        }
        max_filled = loc;
        if (length > max_len) {
            max_len = length;
        }
    }
    return max_len;
}

int getMax(int *a) {
    int max = 0;
    for (int i = 0; i < NRUNS; i++) {
        if (a[i] > max) {
            max = a[i];
        }
    }
    return max;
}

int getMin(int *a) {
    int min = 10000000;
    for (int i = 0; i < NRUNS; i++) {
        if (a[i] < min) {
            min = a[i];
        }
    }
    return min;
}
